Implementation of the fork api for the Workflow.

sagarkapare opened the pull request:
API for the fork looks like following:
```
addFork().addMapReduce("someMapReduce").addSpark("someSpark")
                 .also().addAction(new DummyAction()).addFork()
                                                                               .addMapReduce("anotherMapReduce")
                                                                               .also().addMapReduce("someAnotherMapReduce")
                                                                               .join().addSpark("nextSpark").
                 .also().addAction(new AnotherDummyAction())
                 .join()
```
This will add a fork to the Workflow with 3 branches. First branch will have `someMapReduce` and `someSpark` programs which will be executed sequentially. Second branch will have `DummyAction` after which it forks to 2 branches each executing one map reduce program. After the fork second branch will execute `nextSpark`. Third branch has one `AnotherDummyAction`.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/schedule/SchedulableProgramType.java:
```diff
@@ -20,5 +20,5 @@
  */
 public enum SchedulableProgramType {
-  WORKFLOW, MAPREDUCE, SPARK, CUSTOM_ACTION
+  WORKFLOW, MAPREDUCE, SPARK, CUSTOM_ACTION, NONE
```
Can you explain the `NONE` type a bit?

sagarkapare replied:
`WorkflowSpecification` now have list of `WorkflowNode` instead of list of `ScheduleProgramInfo`. Each `WorkflowNode` in the list has `WorkflowNodeType` to identify type of the node (`ACTION`, `FORK`, `CONDITION`) and the `ScheduleProgramInfo` to identify the program associated with the `WorkflowNode`. In case of `ACTION` nodes, there will be some type of program associated with the WorkflowNode, however `FORK` and `CONDITION` type of WorkflowNode have no program associated with it, which is represented by `NONE` in the above enum. May be we can change the name of the enum to something similar to `WorkflowNodeProgramType` to make it more clear?

chtyim replied:
Isn't the program info is only need if the node type is for running program?

sagarkapare replied:
Yes. Thats true. The implementation can be modified as - `WorkflowNode` can have `name` and `WorkflowNodeType`. Maintain the `map<String, ScheduleProgramInfo> programs` in the `WorkflowSpecification`. If `WorkflowNode` is of type `ACTION`, then corresponding `ScheduleProgramInfo` can be fetched from the `programs` map using `name`. Please let me know if it is more clear.

sagarkapare replied:
@chtyim Removed the need of `NONE` from `ScheduleProgramType`. Please review again. Thanks!

anew commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/AbstractWorkflow.java:
```diff
@@ -106,4 +106,12 @@ protected final void addMapReduce(String mapReduce) {
+   * @return the {@link WorkflowForkConfigurer}
+   */
+  protected final WorkflowForkConfigurer addFork() {
+    return configurer.addFork();
```
I am not a fan of "addFork". Perhaps just fork()...also()...join()?

chtyim replied:
Just call it `fork`

anew commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowConfigurer.java:
```diff
@@ -71,4 +71,10 @@
   void addAction(WorkflowAction action);
+
+  /**
+   * Adds a fork as a next sequential step in the {@link Workflow}
```
Isn't fork parallel? Or is it sequential?

sagarkapare replied:
fork is executed in parallel but its the next sequential node in the `Workflow`. may be will clean documentation a bit.

anew commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowSpecification.java:
```diff
@@ -32,21 +34,28 @@
     this.description = description;
     this.properties = properties == null ? Collections.<String, String>emptyMap() :
                                            Collections.unmodifiableMap(new HashMap<String, String>(properties));
-    this.actions = Collections.unmodifiableList(new ArrayList<ScheduleProgramInfo>(actions));
```
So, if I understand correctly, the workflow is represented as list of nodes. Each node has an id and a type. And for each node type there is a map from id to what the actual node is. Correct? 

Is that natural? The JSON for this workflow spec would be hard to read... 
I could also see this:

- workflow is a list of nodes
- node is an abstract class which provides the node id
- different subclasses of node implement he different types. For example, a fork node has getBranches() which returns a list of list of nodes. 

Wouldn't that be more natural? It keeps everything in one nested structure that reflects the nested structure of the workflow. Or is there a good reason for doing it the way that you did it?

chtyim replied:
Having the class organization based on limitation serialization seems not a good practice.

sagarkapare replied:
@chtyim @anew I have restructured the specifications such that - 
1. `Workflow` contains list of `WorkflowNode`
2. `WorkflowNode` is an abstract class which contains `nodeId` and `nodeType`. 
3. `WorkflowNode` is extended by `WorkflowActionNode` which represents the executable node in the `Workflow` and contains `ScheduleProgramInfo`
4. `WorkflowNode` is also extended by `WorkflowForkNode` which represents fork in the `Workflow` and contain list of `WorkflowForkBranch`. Each `WorkflowForkBranch` simply contains the list of `WorkflowNode`.
5. Custom serialization codecs are added for the `WorkflowNode` and `WorkflowForkBranch` to help serializing and deserializing the `WorkflowNode` based on its type.

anew commented:
code is clean but it seems that the representation of a workflow as a spec is not natural. Please se my other comment.

anew commented:
perhaps get a second opinion from @chtyim

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowForkConfigurer.java:
```diff
@@ -0,0 +1,61 @@
+
+  /**
+   * Joins the current fork
+   * @return the configurer for the parent fork
```
what happen if the parent is not a fork?

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowForkSpecification.java:
```diff
@@ -0,0 +1,41 @@
+  private final List<List<WorkflowNode>> forkBranches;
+
+  public WorkflowForkSpecification(List<List<WorkflowNode>> forkBranches) {
+    this.forkBranches = Collections.unmodifiableList(new ArrayList<List<WorkflowNode>>(forkBranches));
```
Do you need deep-copy of the list?

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowNode.java:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Represents node in the {@link Workflow}.
+ */
+public final class WorkflowNode {
```
So with this, we have `WorkflowActionSpecification`, `ScheduleProgramInfo` and `WorkflowNode`. Shouldn't they got unified into one (or at least in the same hierarchy)?

sagarkapare commented:
@anew @chtyim Please take a look again. Thank you!

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowForkConfigurer.java:
```diff
@@ -0,0 +1,62 @@
+
+/**
+ * Defines an interface for the fork in the {@link Workflow}.
+ * @param <T>
```
Javadoc for `<T>`

sagarkapare replied:
Added.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/AbstractWorkflow.java:
```diff
@@ -106,4 +106,12 @@ protected final void addMapReduce(String mapReduce) {
   }
+
+  /**
+   * Adds a fork to the {@link Workflow}
```
Full stop for the first sentence.

sagarkapare replied:
I think we only add full stop if the comment is for class, or it is for method and has multiple lines.

chtyim replied:
Looks like we turned off the checkstyle for that. But it's recommended (http://checkstyle.sourceforge.net/config_javadoc.html#JavadocStyle and http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#firstsentence)

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/AbstractWorkflow.java:
```diff
@@ -106,4 +106,12 @@ protected final void addMapReduce(String mapReduce) {
+
+  /**
+   * Adds a fork to the {@link Workflow}
+   * @return the {@link WorkflowForkConfigurer}
```
Should be a bit more descriptive. Something like `a {@link WorkflowForkConfigurer} to configure actions in the fork`

sagarkapare replied:
Updated.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowNode.java:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
Why define `equals` and `hashCode` in the abstract base class? Usually it's not a good practice to do so.

sagarkapare replied:
Removed.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowSpecification.java:
```diff
@@ -32,21 +34,22 @@
   private final Map<String, String> properties;
 
+  private List<WorkflowNode> nodes = Lists.newArrayList();
+  private Map<String, WorkflowActionSpecification> customActionMap = Maps.newHashMap();
```
Why we still need this map? Can't it be represented by subclass of `WorkflowNode`? Or you can have `WorkflowActionNode` to be able to carry either `ScheduleProgramInfo` or `WorkflowActionSpecification` for custom action?

sagarkapare replied:
Moved it to `WorkflowActionNode`

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowForkBranch.java:
```diff
@@ -0,0 +1,38 @@
+ * Represents single branch on the {@link WorkflowForkNode}.
+ */
+public class WorkflowForkBranch {
+  private List<WorkflowNode> branch;
```
better call `nodes` or `actions`.

sagarkapare replied:
Updated.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/WorkflowForkBranchCodec.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ *
+ */
+final class WorkflowForkBranchCodec extends AbstractSpecificationCodec<WorkflowForkBranch> {
```
Why need this codec? I think Gson can handle `WorkflowForkBranch` since it's just a POJO with a list inside.

sagarkapare replied:
Removed the codec.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/WorkflowNodeCodec.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.List;
+
+/**
+ *
```
empty javadoc.

sagarkapare replied:
Added javadoc.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/VerificationStage.java:
```diff
@@ -173,6 +155,66 @@ public void process(ApplicationDeployable input) throws Exception {
+        break;
+      case FORK:
+        WorkflowForkNode forkNode = (WorkflowForkNode) node;
+        if (forkNode.getBranches().size() == 0) {
```
This should be a warning in IntelliJ. You should use `isEmpty()` instead of `size() == 0`.

chtyim replied:
Also, you can use `Preconditions.checkArgument` to simplify the code.

sagarkapare replied:
Added `Preconditions.checkArgument`

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/VerificationStage.java:
```diff
@@ -173,6 +155,66 @@ public void process(ApplicationDeployable input) throws Exception {
+                                                   workflowSpec.getName()));
+        }
+
+        for (WorkflowForkBranch branch : forkNode.getBranches()) {
```
To make switch looks nicer and be parity with the `ACTION` case, better encapsulate this into a method `verifyWorkflowFork`

chtyim replied:
In fact, I would move the precondition check into that method as well.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/VerificationStage.java:
```diff
@@ -173,6 +155,66 @@ public void process(ApplicationDeployable input) throws Exception {
+  private void verifyWorkflowAction(ApplicationSpecification appSpec, ScheduleProgramInfo program) {
+    switch (program.getProgramType()) {
+      case MAPREDUCE:
+        if (!appSpec.getMapReduce().containsKey(program.getProgramName())) {
```
Use Preconditions to simplify the code.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java:
```diff
@@ -115,49 +117,61 @@ protected void run() throws Exception {
+      WorkflowNode node = iterator.next();
+      switch (node.getType()) {
+        case ACTION:
+          WorkflowActionNode actionNode = (WorkflowActionNode) node;
```
I don't like big switch case like this. Better encapsulate this action to be taken into a method. e.g.

```java
switch (node.getType()) {
  case ACTION:
    executeAction((WorkflowActionNode) node);
  case FORK:
    executeFork((WorkflowForkNode) node);
  ....
}
```

chtyim replied:
One problem with big switch case is that it's error prone (unintentional fall-through, variables defined in case are not local scope).

sagarkapare replied:
There are some related changes already done in the PR #1386. Since that PR is touching this part of code majorly, will make sure to make these changes in that PR.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowConfigurer.java:
```diff
@@ -66,28 +74,63 @@ public void setProperties(Map<String, String> properties) {
     this.properties = ImmutableMap.copyOf(properties);
   }
 
+  WorkflowNode getWorkflowActionNode(String programName, SchedulableProgramType programType) {
```
Not `get`, but `create`.

sagarkapare replied:
Updated.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowConfigurer.java:
```diff
@@ -66,28 +74,63 @@ public void setProperties(Map<String, String> properties) {
+    return new WorkflowActionNode(nodeId, new ScheduleProgramInfo(programType, programName));
+  }
+
+  WorkflowNode getWorkflowCustomActionNode(WorkflowAction action) {
```
Same here, should be `create`.

sagarkapare replied:
Changed.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowForkConfigurer.java:
```diff
@@ -0,0 +1,107 @@
+  private List<WorkflowNode> currentBranch;
+
+  public DefaultWorkflowForkConfigurer(WorkflowConfigurer workflowConfigurer,
+                                       @Nullable WorkflowForkConfigurer<T> parentForkConfigurer,
```
This should be `T`, isn't it?

sagarkapare replied:
Yes it should be `T`

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowForkConfigurer.java:
```diff
@@ -0,0 +1,107 @@
+ */
+public class DefaultWorkflowForkConfigurer<T> implements WorkflowForkConfigurer<T> {
+
+  private final WorkflowForkConfigurer<T> parentForkConfigurer;
```
This should be of type `T`.

sagarkapare replied:
Changed it.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowForkConfigurer.java:
```diff
@@ -0,0 +1,107 @@
+  @Override
+  @SuppressWarnings("unchecked")
+  public WorkflowForkConfigurer<WorkflowForkConfigurer<T>> fork() {
+    String forkNodeId = ((DefaultWorkflowConfigurer) workflowConfigurer).getNodeIdProvider().getUniqueNodeId();
```
Seems like currently the nodeId are assigned in depth first order for fork branches. Would it better if you generate the node Id in breadth first order and only generate it after the whole workflow graph is finalized (meaning when the `configure()` returns, which in that case, you don't need the `nodeIdProvider`)?

sagarkapare replied:
Currently nodes are assigned in dfs. Meaning all the nodes in the fork will have id's smaller than the id of the node which is next to fork, which seems more appropriate right?

chtyim replied:
But if you look at the node assign of the whole tree, it's neither DFS nor BFS.

sagarkapare replied:
Creating the WorkflowNodes with ids while creating the WorkflowSpecifications now.

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/GoodWorkflowApp.java:
```diff
@@ -0,0 +1,100 @@
+        .fork()
+        .addAction(new DummyAction())
+        .fork()
+        .fork()
```
Please use indentation to indicate nest fork.

chtyim commented:
Done review. Please see my comments.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/ScheduleProgramInfo.java:
```diff
@@ -66,4 +66,13 @@ public int hashCode() {
+
+  @Override
+  public String toString() {
+    final StringBuilder sb = new StringBuilder("ScheduleProgramInfo{");
```
Remove final

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowActionNode.java:
```diff
@@ -0,0 +1,68 @@
+
+  @Override
+  public String toString() {
+    final StringBuilder sb = new StringBuilder("WorkflowActionNode{");
```
Remove final.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowActionNode.java:
```diff
@@ -0,0 +1,68 @@
+   * @return the program information associated with the {@link WorkflowNode}
+   */
+  public ScheduleProgramInfo getProgram() {
+    return this.program;
```
return program

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowForkNode.java:
```diff
@@ -0,0 +1,44 @@
+
+  @Override
+  public String toString() {
+    final StringBuilder sb = new StringBuilder("WorkflowForkNode{");
```
remove final.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowNodeType.java:
```diff
@@ -0,0 +1,26 @@
+public enum WorkflowNodeType {
+  ACTION,
+  FORK,
+  CONDITION
```
If condition is not yet supported, better not to have it now.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/WorkflowNodeCodec.java:
```diff
@@ -0,0 +1,120 @@
+  private void serializeActionNode(WorkflowNode node, JsonObject jsonObj, JsonSerializationContext context) {
+    WorkflowActionNode actionNode = (WorkflowActionNode) node;
+    jsonObj.add("program", context.serialize(actionNode.getProgram(), ScheduleProgramInfo.class));
+    if (actionNode.getProgram().getProgramType() == SchedulableProgramType.CUSTOM_ACTION) {
```
Why need this condition? I thought either program or action spec is null, right? Serializing null is ok (gson will ignore the field).

chtyim replied:
My bad, seems like only spec can be null. Anyway, I think the serialization of any `WorkflowNode` can simply be `context.serialize(node, node.getClass())`, as all sub-types are POJO.

chtyim replied:
So, basically you only need that one line in the `serialize` method above.

sagarkapare replied:
Updated the code per our discussion.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/WorkflowNodeCodec.java:
```diff
@@ -0,0 +1,120 @@
+    WorkflowNode node = null;
+    switch (type) {
+      case ACTION:
+        node = createActionNode(nodeId, jsonObj, context);
```
No need to have separate methods. After you know the `type` (not even need the `nodeId`), do:

```java
switch (type) {
  case ACTION:
    return context.deserialize(json, WorkflowActionNode.class);
  case FORK:
    return context.deserialize(json, WorkflowForkNode.class);
}
```

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowForkConfigurer.java:
```diff
@@ -0,0 +1,100 @@
+
+  @Override
+  public WorkflowForkConfigurer<T> addMapReduce(String mapReduce) {
+    currentBranch.add(((DefaultWorkflowConfigurer) workflowConfigurer).createWorkflowActionNode
```
I don't like this complicated casing and also the need to pass in the top level configurer to just serve as a node factory. Maybe you can have a small package only helper class to create `WorkflowNode` of different types.

sagarkapare replied:
I will still need the top level `DefaultWorkflowConfigurer` because when we do the join for the top level fork, we need to add that fork node to the list of nodes defined inside the `DefaultWorkflowConfigurer`, right?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowForkConfigurer.java:
```diff
@@ -0,0 +1,100 @@
+  @SuppressWarnings("unchecked")
+  public WorkflowForkConfigurer<WorkflowForkConfigurer<T>> fork() {
+    return new DefaultWorkflowForkConfigurer<WorkflowForkConfigurer<T>>
+        (workflowConfigurer, (WorkflowForkConfigurer<T>) this);
```
Why need casting? `this` is already a `WorkflowForkConfigurer<T>`

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowForkConfigurer.java:
```diff
@@ -0,0 +1,100 @@
+    return this;
+  }
+
+  public void addWorkflowForkNode(List<List<WorkflowNode>> branch) {
```
Should be a private method as it's only called from this class.

chtyim commented:
Just couple comments left.

sagarkapare commented:
@chtyim Addressed the comments. Please take a look again. Thanks!

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/WorkflowNodeCreator.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Helper to create {@link WorkflowNode}
+ */
+class WorkflowNodeCreator {
```
final class, also have a private Constructor. That's the standard for util/helper class.

chtyim commented:
Just one comment left.

I do have a higher level question. Does the codec be able to deserialize older version of workflow spec of a deployed app?

sagarkapare commented:
@chtyim Addressed comment. Please take a look again!

Regarding deserialization of the older version of workflow spec, I have created blocker jira - https://issues.cask.co/browse/CDAP-1621 and will address the issue in separate PR. 
Possible approach is to have `version` field in the `WorkflowSpecification`. Since this field will be absent in the old app workflow spec, while deserializing `WorkflowSpecificationCodec` can determine if the spec are for older app and based on that it can read and migrate those spec to the new version.

sagarkapare commented:
Build is running here: http://builds.cask.co/browse/CDAP-DUT777

chtyim commented:
LGTM as long as you resolve the conflicts correctly :)

sagarkapare commented:
Build passed - http://builds.cask.co/browse/CDAP-DUT777-11 after resolving conflicts :) Merging.
